File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/LogHeadRecord.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System.Fabric;
    using System.IO;

    internal abstract class LogHeadRecord : PhysicalLogRecord
    {
        private static readonly uint ApproximateDiskSpaceUsed = sizeof(int) + sizeof(ulong) + (4*sizeof(long));

        private Epoch logHeadEpoch;

        private LogicalSequenceNumber logHeadLsn;

        private PhysicalSequenceNumber logHeadPsn;

        // The following fields are not persisted
        private IndexingLogRecord logHeadRecord;

        private ulong logHeadRecordOffset;

        internal LogHeadRecord(LogRecordType recordType, ulong recordPosition, long lsn)
            : base(recordType, recordPosition, lsn)
        {
            this.logHeadEpoch = LogicalSequenceNumber.InvalidEpoch;
            this.logHeadLsn = LogicalSequenceNumber.InvalidLsn;
            this.logHeadPsn = PhysicalSequenceNumber.InvalidPsn;
            this.logHeadRecordOffset = InvalidPhysicalRecordOffset;
            this.logHeadRecord = IndexingLogRecord.InvalidIndexingLogRecord;
        }

        internal LogHeadRecord(
            LogRecordType recordType,
            IndexingLogRecord logHeadRecord,
            LogicalSequenceNumber lsn,
            PhysicalLogRecord lastLinkedPhysicalRecord)
            : base(recordType, lsn, lastLinkedPhysicalRecord)
        {
            this.logHeadEpoch = logHeadRecord.CurrentEpoch;
            this.logHeadLsn = logHeadRecord.Lsn;
            this.logHeadPsn = logHeadRecord.Psn;
            this.logHeadRecordOffset = InvalidPhysicalRecordOffset;
            this.logHeadRecord = logHeadRecord;

            this.UpdateApproximateDiskSize();
        }

        internal IndexingLogRecord HeadRecord
        {
            get { return this.logHeadRecord; }

            set { this.logHeadRecord = value; }
        }

        internal Epoch LogHeadEpoch
        {
            get { return this.logHeadEpoch; }
        }

        internal LogicalSequenceNumber LogHeadLsn
        {
            get { return this.logHeadLsn; }
        }

        internal PhysicalSequenceNumber LogHeadPsn
        {
            get { return this.logHeadPsn; }
        }

        internal ulong LogHeadRecordOffset
        {
            get { return this.logHeadRecordOffset; }
        }

        internal ulong LogHeadRecordPosition
        {
            get { return this.RecordPosition - this.logHeadRecordOffset; }
        }

        internal override bool FreePreviousLinksLowerThanPsn(PhysicalSequenceNumber newHeadPsn)
        {
            bool ret = base.FreePreviousLinksLowerThanPsn(newHeadPsn);

            if (this.logHeadRecord != null &&
                this.logHeadRecord.Psn < newHeadPsn)
            {
                Utility.Assert(
                    this.logHeadRecordOffset != InvalidPhysicalRecordOffset,
                    "this.logHeadRecordOffset must be valid");

                this.logHeadRecord = 
                    (this.logHeadRecordOffset == 0) ? 
                        null : 
                        IndexingLogRecord.InvalidIndexingLogRecord;

                return true;
            }

            return ret;
        }

        protected override void Read(BinaryReader br, bool isPhysicalRead)
        {
            base.Read(br, isPhysicalRead);

            var startingPosition = br.BaseStream.Position;
            var sizeOfSection = br.ReadInt32();
            var endPosition = startingPosition + sizeOfSection;

            var dataLossNumber = br.ReadInt64();
            var configurationNumber = br.ReadInt64();
            this.logHeadEpoch = new Epoch(dataLossNumber, configurationNumber);
            this.logHeadLsn = new LogicalSequenceNumber(br.ReadInt64());
            this.logHeadPsn = new PhysicalSequenceNumber(br.ReadInt64());
            if (isPhysicalRead == true)
            {
                this.logHeadRecordOffset = br.ReadUInt64();
            }

            // Jump to the end of the section ignoring fields that are not understood.
            Utility.Assert(endPosition >= br.BaseStream.Position, "Could not have read more than section size.");
            br.BaseStream.Position = endPosition;

            this.UpdateApproximateDiskSize();
            return;
        }

        protected override void Write(BinaryWriter bw, OperationData operationData, bool isPhysicalWrite, bool forceRecomputeOffsets)
        {
            base.Write(bw, operationData, isPhysicalWrite, forceRecomputeOffsets);
            var startingPos = bw.BaseStream.Position;
            bw.BaseStream.Position += sizeof(int);

            bw.Write(this.logHeadEpoch.DataLossNumber);
            bw.Write(this.logHeadEpoch.ConfigurationNumber);
            bw.Write(this.logHeadLsn.LSN);
            bw.Write(this.logHeadPsn.PSN);

            if (isPhysicalWrite == true)
            {
                if (this.logHeadRecordOffset == InvalidPhysicalRecordOffset || forceRecomputeOffsets == true)
                {
                    Utility.Assert(
                        this.logHeadRecord != null,
                        "this.logHeadRecord != null");
                    Utility.Assert(
                        this.logHeadRecord.RecordPosition != InvalidRecordPosition,
                        "this.logHeadRecord.RecordPosition != LogRecord.INVALID_RECORD_POSITION");
                    Utility.Assert(
                        this.RecordPosition != InvalidRecordPosition,
                        "this.RecordPosition != LogRecord.INVALID_RECORD_POSITION");
                    this.logHeadRecordOffset = this.RecordPosition - this.logHeadRecord.RecordPosition;
                }

                bw.Write(this.logHeadRecordOffset);
            }

            var endPosition = bw.BaseStream.Position;
            var sizeOfSection = checked((int) (endPosition - startingPos));
            bw.BaseStream.Position = startingPos;
            bw.Write(sizeOfSection);
            bw.BaseStream.Position = endPosition;

            operationData.Add(CreateArraySegment(startingPos, bw));
            return;
        }

        private void UpdateApproximateDiskSize()
        {
            this.ApproximateSizeOnDisk += ApproximateDiskSpaceUsed;
        }
    }
}

File: src/prod/src/data/txnreplicator/logrecordlib/LogHeadRecord.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LogRecordLib;
using namespace TxnReplicator;
using namespace Data::Utilities;

const ULONG LogHeadRecord::DiskSpaceUsed = sizeof(LONG32) + sizeof(ULONG64) + (4 * sizeof(LONG64));

LogHeadRecord::LogHeadRecord(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in IndexingLogRecord & invalidIndexingLogRecord)
    : PhysicalLogRecord(recordType, recordPosition, lsn, invalidPhysicalLogRecord)
    , logHeadLsn_(Constants::InvalidLsn)
    , logHeadPsn_(Constants::InvalidPsn)
    , logHeadRecordOffset_(Constants::InvalidPhysicalRecordOffset)
    , logHeadEpoch_(Epoch::InvalidEpoch())
    , logHeadRecord_(&invalidIndexingLogRecord)
{
}

LogHeadRecord::LogHeadRecord(
    __in LogRecordType::Enum recordType,
    __in IndexingLogRecord & logHeadRecord,
    __in LONG64 lsn,
    __in_opt PhysicalLogRecord * const lastLinkedPhysicalRecord,
    __in PhysicalLogRecord & invalidPhysicalRecord)
    : PhysicalLogRecord(recordType, lsn, lastLinkedPhysicalRecord, invalidPhysicalRecord)
    , logHeadLsn_(logHeadRecord.Lsn)
    , logHeadPsn_(logHeadRecord.Psn)
    , logHeadRecordOffset_(Constants::InvalidPhysicalRecordOffset)
    , logHeadEpoch_(logHeadRecord.CurrentEpoch)
    , logHeadRecord_(&logHeadRecord)
{
    UpdateApproximateDiskSize();
}

LogHeadRecord::~LogHeadRecord()
{
}

void LogHeadRecord::UpdateApproximateDiskSize()
{
    ApproximateSizeOnDisk = ApproximateSizeOnDisk + DiskSpaceUsed;
}

void LogHeadRecord::Read(
    __in BinaryReader & binaryReader,
    __in bool isPhysicalRead)
{
    __super::Read(binaryReader, isPhysicalRead);

    ULONG32 startingPosition = binaryReader.Position;
    ULONG32 sizeOfSection = 0;
    binaryReader.Read(sizeOfSection);

    ULONG32 endPosition = startingPosition + sizeOfSection;

    // Read signed LONGLONG - Int64 in Managed
    LONG64 dataLossNumber;
    LONG64 configurationNumber;
    binaryReader.Read(dataLossNumber);
    binaryReader.Read(configurationNumber);

    logHeadEpoch_ = Epoch(dataLossNumber, configurationNumber);

    binaryReader.Read(logHeadLsn_);
    binaryReader.Read(logHeadPsn_);

    if(isPhysicalRead)
    {
        binaryReader.Read(logHeadRecordOffset_);
    }

    ASSERT_IFNOT(
        endPosition >= binaryReader.Position,
        "Invalid log head record, end position: {0} reader position:{1}",
        endPosition, 
        binaryReader.Position);

    binaryReader.Position = endPosition;

    UpdateApproximateDiskSize();
}

void LogHeadRecord::Write(
    __in BinaryWriter & binaryWriter,
    __inout OperationData & operationData,
    __in bool isPhysicalWrite,
    __in bool forceRecomputeOffsets)
{
    __super::Write(binaryWriter, operationData, isPhysicalWrite, forceRecomputeOffsets);
    ULONG32 startingPosition = binaryWriter.Position;

    binaryWriter.Position += sizeof(int);

    binaryWriter.Write(logHeadEpoch_.DataLossVersion);
    binaryWriter.Write(logHeadEpoch_.ConfigurationVersion);
    binaryWriter.Write(logHeadLsn_);
    binaryWriter.Write(logHeadPsn_);

    if(isPhysicalWrite)
    {
        if(logHeadRecordOffset_ == Constants::InvalidPhysicalRecordOffset || forceRecomputeOffsets == true)
        {
            ASSERT_IFNOT(
                logHeadRecord_ != nullptr,
                "logHeadRecord_ must not be null");

            ASSERT_IFNOT(
                logHeadRecord_->RecordPosition != Constants::InvalidRecordPosition,
                "logHeadRecord_->RecordPosition must be valid");

            ASSERT_IFNOT(
                RecordPosition != Constants::InvalidRecordPosition,
                "RecordPosition must be valid");

            logHeadRecordOffset_ = RecordPosition - logHeadRecord_->RecordPosition;
        }

        binaryWriter.Write(logHeadRecordOffset_);
    }

    ULONG32 endPosition = binaryWriter.Position;
    ULONG32 sizeOfSection = endPosition - startingPosition;

    binaryWriter.Position = startingPosition;
    binaryWriter.Write(sizeOfSection);
    binaryWriter.Position = endPosition;

    KBuffer::SPtr kBuffer = binaryWriter.GetBuffer(startingPosition);
    operationData.Append(*kBuffer);
}

bool LogHeadRecord::FreePreviousLinksLowerThanPsn(
    __in LONG64 newHeadPsn,
    __in InvalidLogRecords & invalidLogRecords)
{
    bool ret = __super::FreePreviousLinksLowerThanPsn(
        newHeadPsn,
        invalidLogRecords);

    if (logHeadRecord_ != nullptr &&
        logHeadRecord_->Psn < newHeadPsn)
    {
        ASSERT_IF(
            logHeadRecordOffset_ == Constants::InvalidPhysicalRecordOffset,
            "logHeadRecordOffset_ must be valid");

        logHeadRecord_ = 
            logHeadRecordOffset_ == 0 ?
            nullptr :
            invalidLogRecords.Inv_IndexingLogRecord;
        return true;
    }

    return ret;
}

bool LogHeadRecord::Test_Equals(__in LogRecord const & other) const
{
    LogHeadRecord const & otherLogHeadRecord = dynamic_cast<LogHeadRecord const &>(other);

    if (__super::Test_Equals(other))
    {
        return
            logHeadLsn_ == otherLogHeadRecord.logHeadLsn_ &&
            logHeadPsn_ == otherLogHeadRecord.logHeadPsn_ &&
            logHeadRecordOffset_ == otherLogHeadRecord.logHeadRecordOffset_ &&
            logHeadEpoch_ == otherLogHeadRecord.logHeadEpoch_;
    }

    return false;
}
